### PR TITLE
Split `module.bc.zstd` into `root.bc.zstd` and `functions.bc.zstd`.

### PR DESCRIPTION
**Pull request by Cvelth**

### CODE REVIEW

File: lib/PipelineC/PipelineC.cpp [hunk 1]
```diff
@@ -69,6 +69,8 @@ class ExistingOrNew {
 
 static void llvmErrorToRpError(llvm::Error Error, rp_error *Out) {
   if (Out == nullptr) {
+    // TODO: this is suspicions, isn't it an indication that callee is
+    // doing something nasty? Is it ok to just ignore it?
```
Comment (mrjackv) [inline lib/PipelineC/PipelineC.cpp]: That's part of the `PipelineC` calling conventions, if you see the callers of this functions (e.g. `_rp_manager_run_analysis`), they pass through the `rp_error *` variable. The callee of these can opt-in to receive errors by passing the `rp_error *`, whereas they get ignored if `NULL` is provided as an argument. (Essentially the `rp_error *` constitutes an optional return value, but due to C structure that was one of the saner ways to implement it).
There's a section `Error Reporting` in `PipelineC/Prototypes.h` explaining this in depth.

File: lib/PipelineC/PipelineC.cpp [hunk 2]
```diff
@@ -93,6 +95,8 @@ static void llvmErrorToRpError(llvm::Error Error, rp_error *Out) {
     *Out = rp_simple_error(Reason, "");
   };
 
+  // TODO: do we really need `handleAllErrors` here? It makes it seem as if
+  // we're overriding `*Out` a bunch of times, even though it only happens once.
```
Comment (mrjackv) [inline lib/PipelineC/PipelineC.cpp]: The `All` here means that if an error is unhandled it aborts. We're guaranteed that that doesn't happen because the second handler accepts the base class `llvm::ErrorInfoBase`.

File: lib/Pipeline/Context.cpp [hunk 3]
```diff
@@ -44,6 +44,7 @@ llvm::Error Context::load(const revng::DirectoryPath &Path) {
   if (not MaybeExists)
     return MaybeExists.takeError();
 
+  // TODO: this is suspicious, we should probably add a comment why.
```
Comment (mrjackv) [inline lib/Pipeline/Context.cpp]: The check below is basically `stat`, so if the file doesn't exist we stop and succeed. Either way a comment doesn't hurt.

File: share/revng/test/tests/daemon/test.py [hunk 4]
```diff
@@ -129,9 +129,9 @@ def error_handler(e: BaseException):
 
     binary = pytestconfig.getoption("binary")
     tracing = TraceConfig()
-    tracing.on_request_headers_sent.append(header_trace)
-    tracing.on_request_chunk_sent.append(payload_trace)
-    tracing.on_request_end.append(response_trace)
+    tracing.on_request_headers_sent.append(header_trace)  # type: ignore
+    tracing.on_request_chunk_sent.append(payload_trace)  # type: ignore
+    tracing.on_request_end.append(response_trace)  # type: ignore
```
Comment (mrjackv) [inline share/revng/test/tests/daemon/test.py]: This commit is no longer required, this was fixed by the `feature/pypeline` branch, which has been merged

File: lib/Pipeline/Context.cpp (hunk 3 above)
Comment (Cvelth) [reply lib/Pipeline/Context.cpp]: Yeah, that's what I mean, it's suspicious for a _load_ function to succeed when the file it's loading _from_ doesn't exist.

File: share/revng/test/tests/daemon/test.py (hunk 4 above)
Comment (Cvelth) [reply share/revng/test/tests/daemon/test.py]: Ah, my bad, I didn't mean to push this! Thanks for catching this!

File: lib/PipelineC/PipelineC.cpp (hunk 1 above)
Comment (Cvelth) [reply lib/PipelineC/PipelineC.cpp]: Thanks for the explanation! I'll add a comment pointing to `PipelineC/Prototypes.h`

File: lib/Pipeline/Context.cpp (hunk 3 above)
Comment (mrjackv) [reply lib/Pipeline/Context.cpp]: It makes sense in the context of `lib/Pipeline`, both context and containers are default constructible, so the logic is:
```c
Context C;
if (exists(context_file))
  C.deserialize(context_file.read());
 return C;
 ```

Comment (Cvelth) [reply lib/Pipeline/Context.cpp]: I see. I'll add a comment!

File: lib/PipelineC/PipelineC.cpp (hunk 2 above)
Comment (Cvelth) [reply lib/PipelineC/PipelineC.cpp]: Ah, wait, it looks like a misunderstanding of mine!

For some reason I assumed there's also a _single_ error handler that guarantees no additional errors are spawned (and as the result, it doesn't have to be recursive). But it looks like I made it up, my bad, sorry for the noise!

(keeping the thread open so that I don't forget to remove the comment)

File: share/revng/test/tests/daemon/test.py (hunk 4 above)
Comment (Cvelth) [reply share/revng/test/tests/daemon/test.py]: Dropped.

File: lib/PipelineC/PipelineC.cpp (hunk 1 above)
Comment (Cvelth) [reply lib/PipelineC/PipelineC.cpp]: Done:
```
    // By passing `Out == nullptr` the callee indicated that they do not care
    // about the error, so we can just discard it.
    //
    // For more context, see `include/revng/PipelineC/Prototypes.h`.
```

File: lib/PipelineC/PipelineC.cpp (hunk 2 above)
Comment (Cvelth) [reply lib/PipelineC/PipelineC.cpp]: Gone.

File: lib/Pipeline/Context.cpp (hunk 3 above)
Comment (Cvelth) [reply lib/Pipeline/Context.cpp]: Done:
```
  // In the context of `lib/Pipeline`, both context and containers are default
  // constructible. As such, we just skip loading files that do not exist,
  // ending up with the default state.
```

File: lib/Model/Importer/Binary/ELFImporter.cpp [hunk 5]
```diff
@@ -487,8 +487,8 @@ void ELFImporter<T, HasAddend>::findMissingTypes(object::ELFFile<T> &TheELF,
       if (auto E = importELF(DepModel, *TheBinary, AdjustedOptions)) {
         revng_log(ELFImporterLog,
                   "Can't import model for " << DependencyLibrary << " due to "
-                                            << E);
-        llvm::consumeError(std::move(E));
+                                            << Error);
+        llvm::consumeError(std::move(Error));
         ModelsOfLibraries.erase(DependencyLibrary);
         continue;
       }
```
Comment (aleclearmind) [inline lib/Model/Importer/Binary/ELFImporter.cpp]: Missing rename of `E` in declaration?

File: include/revng/ADT/UpcastablePointer.h [hunk 6]
```diff
@@ -64,6 +64,11 @@ ReturnT upcast(P &&Upcastable, const L &Callable, ReturnT &&IfNull) {
   if constexpr (I < std::tuple_size_v<concrete_types>) {
     using type = std::tuple_element_t<I, concrete_types>;
     if (auto *Upcasted = llvm::dyn_cast<type>(Pointer)) {
+      if constexpr (std::same_as<std::decay_t<ReturnT>, llvm::Error>)
+        llvm::consumeError(std::move(IfNull));
+      if constexpr (SpecializationOf<std::decay_t<ReturnT>, llvm::Expected>)
+        llvm::consumeError(std::move(IfNull).takeError());
```
Comment (aleclearmind) [inline include/revng/ADT/UpcastablePointer.h]: I don't think returning `Expected` ever makes sense.

File: include/revng/ADT/UpcastablePointer.h [hunk 7]
```diff
@@ -75,27 +80,6 @@ ReturnT upcast(P &&Upcastable, const L &Callable, ReturnT &&IfNull) {
   }
 }
 
-template<typename L, UpcastablePointerLike P, size_t I = 0>
-llvm::Error upcast(P &&Upcastable, const L &Callable, llvm::Error IfNull) {
```
Comment (aleclearmind) [inline include/revng/ADT/UpcastablePointer.h:79]: Add a Doxygen comment about this use case on the remaining definition of this.

File: lib/Model/Importer/Binary/MachOImporter.cpp [hunk 8]
```diff
@@ -282,6 +282,7 @@ Error MachOImporter::import() {
   }
 
   Error TheError = Error::success();
+  revng_check(TheError);
```
Comment (aleclearmind) [inline lib/Model/Importer/Binary/MachOImporter.cpp]: `o_O`

File: lib/Support/ResourceFinder.cpp [hunk 9]
```diff
@@ -28,8 +28,8 @@ std::string getComponentsHash() {
 
   std::string Result;
   for (std::string &File : Files) {
-    auto MaybeBuffer = llvm::MemoryBuffer::getFile(File);
-    revng_assert(MaybeBuffer);
+    auto MaybeBuffer = errorOrToExpected(llvm::MemoryBuffer::getFile(File));
+    revng_check(MaybeBuffer);
```
Comment (aleclearmind) [inline lib/Support/ResourceFinder.cpp]: `cantFail`?

File: tests/unit/Pipeline.cpp [hunk 10]
```diff
@@ -657,7 +663,8 @@ BOOST_AUTO_TEST_CASE(SingleElementPipelineFailure) {
 
   auto Error = Pipeline.run("end", Targets);
   BOOST_TEST(!!Error);
-  consumeError(std::move(Error));
+  if (!!Error)
```
Comment (aleclearmind) [inline tests/unit/Pipeline.cpp]: The `!!` pattern seems suspicious to me. Can we avoid it or is there a reason?

File: tests/unit/Pipeline.cpp [hunk 11]
```diff
@@ -534,6 +534,9 @@ BOOST_AUTO_TEST_CASE(SingleElementPipelineCanBeRunned) {
   Targets[CName].emplace_back(Target(RootKind2));
   auto Error = Pip.run("end", Targets);
   BOOST_TEST(!Error);
+  if (Error)
+    llvm::consumeError(std::move(Error));
```
Comment (aleclearmind) [inline tests/unit/Pipeline.cpp]: What's the point?

File: tests/unit/Pipeline.cpp [hunk 12]
```diff
@@ -822,9 +833,7 @@ BOOST_AUTO_TEST_CASE(SingleElementPipelineForwardFinedGrained) {
   llvm::StringMap<ContainerToTargetsMap> Invalidations;
   Invalidations[Name].add(CName, {}, RootKind);
 
-  auto Error = Pipeline.getInvalidations(Invalidations);
-  BOOST_TEST(!Error);
-
+  Pipeline.getInvalidations(Invalidations);
```
Comment (aleclearmind) [inline tests/unit/Pipeline.cpp:825]: I think this commit belongs to a previous patch.
I'd squash them to be honest.

File: python/revng/internal/cli/_commands/override_by_name.py [hunk 13]
```diff
@@ -68,7 +68,7 @@ def run(self, options: Options):
 
             # Extract model, if necessary
             result = run_revng_command(
-                ["model", "opt", "-Y", saved_file.name, "-o", model_file.name], options
+                ["model", "opt", saved_file.name, "-o", model_file.name], options
```
Comment (aleclearmind) [inline python/revng/internal/cli/_commands/override_by_name.py:71]: Can we drop it from `model opt` then?

File: lib/ImportFromC/HeaderToModel.cpp [hunk 14]
```diff
@@ -1265,6 +1265,8 @@ void HeaderToModelDiagnosticConsumer::HandleDiagnostic(Level DiagLevel,
   llvm::raw_string_ostream OS(Text);
   auto *DiagOpts = &Info.getDiags()->getDiagnosticOptions();
 
+  // Suspicious!
+  // TODO: investigate.
   uint64_t StartOfLocationInfo = OS.tell();
```
Comment (aleclearmind) [inline lib/ImportFromC/HeaderToModel.cpp]: Use `// WIP` if you don't want to merge.
Anyway, I can't help with this.

File: lib/Model/Importer/Binary/ELFImporter.cpp [hunk 15]
```diff
@@ -224,6 +224,7 @@ Error ELFImporter<T, HasAddend>::import(const ImporterOptions &Options) {
 
       auto MaybeSectionName = TheELF.getSectionName(SectionHeader);
       if (auto Error = MaybeSectionName.takeError()) {
+        // TODO: is there anything useful we can extract from this error to log?
```
Comment (aleclearmind) [inline lib/Model/Importer/Binary/ELFImporter.cpp]: Yes, we should *never* silently suppress errors! Log!

File: lib/Model/Importer/Binary/ELFImporter.cpp [hunk 16]
```diff
@@ -461,6 +462,9 @@ void ELFImporter<T, HasAddend>::findMissingTypes(object::ELFFile<T> &TheELF,
       revng_log(ELFImporterLog, " Importing Model for: " << DependencyLibrary);
       auto BinaryOrErr = llvm::object::createBinary(DependencyLibrary);
       if (auto Error = BinaryOrErr.takeError()) {
+        // TODO: wouldn't user want to know we couldn't find one of
+        //       the dependencies?
+        //       Sounds like we want a better way to report this.
```
Comment (aleclearmind) [inline lib/Model/Importer/Binary/ELFImporter.cpp]: I'd say this is a candidate for diagnostics that can get to the user.

As of now, we don't have it. I suggest to introduce a new convention for logs that should be turned into diagnostic:

```
// TODO: turn into diagnostic
```

File: lib/Pipeline/LLVMContainer.cpp [hunk 17]
```diff
@@ -65,14 +65,14 @@ std::unique_ptr<ContainerBase>
 LLVMContainer::cloneFiltered(const TargetsList &Targets) const {
   using InspectorT = LLVMKind;
   auto ToClone = InspectorT::functions(Targets, *this->self());
-  auto ToClonedNotOwned = InspectorT::untrackedFunctions(*this->self());
+  auto ToCloneUntracted = InspectorT::untrackedFunctions(*this->self());
```
Comment (aleclearmind) [inline lib/Pipeline/LLVMContainer.cpp]: Untracked?

File: include/revng/Pipeline/LLVMContainer.h [hunk 18]
```diff
@@ -46,6 +46,9 @@ class LLVMContainer : public EnumerableContainer<LLVMContainer> {
                 std::unique_ptr<llvm::Module> M) :
     EnumerableContainer<ThisType>(*Context, Name), Module(std::move(M)) {}
 
+  LLVMContainer &cloneFrom(const LLVMContainer &Another);
+  LLVMContainer &swapWith(LLVMContainer &Another);
```
Comment (aleclearmind) [inline include/revng/Pipeline/LLVMContainer.h:50]: I'd squash this commit in the first commit using it.
Same for `linkModules`.

File: share/revng/pipelines/revng-pipelines.yml [hunk 19]
```diff
@@ -148,14 +152,10 @@ Branches:
             Therefore, they lack arguments and return values.
       - Name: enforce-abi
         Pipes:
-          - Type: llvm-pipe
-            UsedContainers: [module.bc.zstd]
-            Passes:
-              - drop-root
```
Comment (aleclearmind) [inline share/revng/pipelines/revng-pipelines.yml:154]: I was like "nononono", but it's "yesyesyes".

We should have done this much sooner.

File: share/revng/pipelines/revng-pipelines.yml [hunk 20]
```diff
@@ -506,15 +506,17 @@ Branches:
...
           - Type: embed-statement-comments
-            UsedContainers: [module.bc.zstd]
+            UsedContainers: [functions.bc.zstd]
       - Name: decompile
         Pipes:
           - Type: helpers-to-header
-            UsedContainers: [module.bc.zstd, helpers.h]
+            UsedContainers: [functions.bc.zstd, helpers.h]
           - Type: model-to-header
+            # **ivan** shouldn't we move this further up the pipeline?
```
Comment (aleclearmind) [inline share/revng/pipelines/revng-pipelines.yml]: Use `WIP` for comments that should not be merged.
Don't let your name get merged.

File: share/revng/pipelines/revng-pipelines.yml [hunk 21]
```diff
@@ -9,7 +9,11 @@ Containers:
   - Name: cross-relations.yml
     Type: binary-cross-relations
     Role: cross-relations
-  - Name: module.bc.zstd
+  - Name: functions.bc.zstd
+    Type: llvm-container
+  - Name: root.bc.zstd
+    Type: llvm-container
+  - Name: recompilable-root.bc.zstd
```
Comment (aleclearmind) [inline share/revng/pipelines/revng-pipelines.yml]: The name seems to imply that `root.bc.zstd` is not recompilable, while it is, on another branch of the pipeline.

Let's go with `root-and-functions.bc.zstd` or `root-with-functions.bc.zstd`.

File: lib/FunctionIsolation/InvokeIsolatedFunctions.cpp [hunk 22]
```diff
@@ -190,15 +163,21 @@ class InvokeIsolatedFunctions {
...
             auto Name = model::Register::getCSVName(Register);
-            GlobalVariable *CSV = M->getGlobalVariable(Name, true);
+            GlobalVariable *CSV = RootModule.getGlobalVariable(Name, true);
             revng_assert(CSV != nullptr);
             Arguments.push_back(createLoad(Builder, CSV));
           }
         }
       }
 
+      llvm::Function *
```
Comment (aleclearmind) [inline lib/FunctionIsolation/InvokeIsolatedFunctions.cpp:172]: You can use `auto *` here

File: lib/FunctionIsolation/InvokeIsolatedFunctions.cpp [hunk 23]
```diff
@@ -23,78 +26,48 @@
...
 
-    for (BasicBlock &BB : *RootFunction) {
+    for (BasicBlock &BB : RootFunction) {
       revng_assert(not BB.empty());
 
       MetaAddress JumpTarget = getBasicBlockJumpTarget(&BB);
       auto It = Map.find(JumpTarget);
       if (It != Map.end()) {
+        revng_assert(get<1>(It->second) == nullptr,
+                     "are we ok with overriding basic blocks here?");
```
Comment (aleclearmind) [inline lib/FunctionIsolation/InvokeIsolatedFunctions.cpp]: Please do not leave questions in asserts, I find it quite confusing.

State your expectations, not questions.

File: lib/FunctionIsolation/InvokeIsolatedFunctions.cpp [hunk 24]
```diff
@@ -190,15 +163,21 @@ class InvokeIsolatedFunctions {
...
-            GlobalVariable *CSV = M->getGlobalVariable(Name, true);
+            GlobalVariable *CSV = RootModule.getGlobalVariable(Name, true);
             revng_assert(CSV != nullptr);
             Arguments.push_back(createLoad(Builder, CSV));
           }
         }
       }
 
+      llvm::Function *
+        NewDeclaration = llvm::Function::Create(F->getFunctionType(),
```
Comment (aleclearmind) [inline lib/FunctionIsolation/InvokeIsolatedFunctions.cpp:173]: Using function types of another module scares me, but it should be fine :thinking:

File: include/revng/ADT/UpcastablePointer.h (hunk 6 above)
Comment (Cvelth) [reply include/revng/ADT/UpcastablePointer.h]: Yes, but better safe than sorry, this does not have any impact and might be helpful if someone does put an expected in there.

That being said, if you prefer, I wouldn't mind
```cpp
static_assert(not SpecializationOf<std::decay_t<ReturnT>, llvm::Expected>);
```
either.

File: lib/Model/Importer/Binary/MachOImporter.cpp (hunk 8 above)
Comment (Cvelth) [reply lib/Model/Importer/Binary/MachOImporter.cpp]: Better safe than sorry :D

Weird copy-paste could lead to an assertion without this.

File: tests/unit/Pipeline.cpp (hunk 10 above)
Comment (Cvelth) [reply tests/unit/Pipeline.cpp]: I just kept it because it was there. That's literally just `bool(Error)` but not as ugly, which could be relevant in some cases. Doesn't matter here though.

File: tests/unit/Pipeline.cpp (hunk 11 above)
Comment (Cvelth) [reply tests/unit/Pipeline.cpp]: I'm not too familiar with Boost.Test in particular but I'd assume there's a way (like a CLI argument or something) to turn these into "print"-like statements instead of aborts or something akin to that.

But yeah, just safety for a corner case, I don't mind dropping it.

File: python/revng/internal/cli/_commands/override_by_name.py (hunk 13 above)
Comment (Cvelth) [reply python/revng/internal/cli/_commands/override_by_name.py:71]: I did in one of the previous commits.

The thing is it was done through a helper in a header, so it was a single action to drop it everywhere.

File: lib/Pipeline/LLVMContainer.cpp (hunk 17 above)
Comment (Cvelth) [reply lib/Pipeline/LLVMContainer.cpp]: For terminology consistency, the function name uses `untracked` prefix.

File: lib/Model/Importer/Binary/ELFImporter.cpp (hunk 5 above)
Comment (Cvelth) [reply lib/Model/Importer/Binary/ELFImporter.cpp]: Good catch! This was commit reordering fiasco, the declaration rename got stuck in a commit that came later.

Fixed.

File: include/revng/ADT/UpcastablePointer.h (hunk 7 above)
Comment (Cvelth) [reply include/revng/ADT/UpcastablePointer.h:79]: Done:
```
/// This is an upcasting helper `UpcastablePointer` below takes advantage of
/// heavily. When possible, prefer the member function instead.
///
/// That being said, this free function is not tied to the object and can be
/// used on normal pointers, as long as LLVM-style RTTI conventions
/// are adhered to.
///
/// \note this also correctly handles `llvm::Error` and `llvm::Expected`
/// when they are passed into \param IfNull.
```

File: lib/Support/ResourceFinder.cpp (hunk 9 above)
Comment (Cvelth) [reply lib/Support/ResourceFinder.cpp]: Done.

File: tests/unit/Pipeline.cpp (hunk 10 above)
Comment (Cvelth) [reply tests/unit/Pipeline.cpp]: Ok, I was wrong, it is essential in `BOOST_TEST`s, as those do not convert the input type to bool AND they pass it as is into `operator<<` of the std-stream, which llvm types don't play nicely with.

I'd suggest a more clean approach to the error being checked and, more importantly, printed properly, but let's be honest, this is old pipeline tests - those go away soon anyway. I say let them be.

File: tests/unit/Pipeline.cpp (hunk 12 above)
Comment (Cvelth) [reply tests/unit/Pipeline.cpp:825]: Squashed

File: lib/ImportFromC/HeaderToModel.cpp (hunk 14 above)
Comment (Cvelth) [reply lib/ImportFromC/HeaderToModel.cpp]: I looked into this, and yes, this is clang infrastructure being awkward, not our fault.

Closing this.

File: lib/Model/Importer/Binary/ELFImporter.cpp (hunk 16 above)
Comment (Cvelth) [reply lib/Model/Importer/Binary/ELFImporter.cpp]: I went with
```
        // TODO: emit a diagnostic message for the user.
```

File: include/revng/Pipeline/LLVMContainer.h (hunk 18 above)
Comment (Cvelth) [reply include/revng/Pipeline/LLVMContainer.h:50]: Squashed.

File: share/revng/pipelines/revng-pipelines.yml (hunk 20 above)
Comment (Cvelth) [reply share/revng/pipelines/revng-pipelines.yml]: Whoops, this one also looks like rebasing fiasco, my bad and good catch.

Comment (Cvelth) [reply share/revng/pipelines/revng-pipelines.yml]: (as for the comment itself, I got rid of it - it's for the old pipeline anyway, who cares, and I'll be the one adding the new equivalent of these pipes, so...)

File: share/revng/pipelines/revng-pipelines.yml (hunk 21 above)
Comment (Cvelth) [reply share/revng/pipelines/revng-pipelines.yml]: Done (`with` one).

File: lib/FunctionIsolation/InvokeIsolatedFunctions.cpp (hunk 22 above)
Comment (Cvelth) [reply lib/FunctionIsolation/InvokeIsolatedFunctions.cpp:172]: It doesn't change anything.

File: lib/FunctionIsolation/InvokeIsolatedFunctions.cpp (hunk 23 above)
Comment (Cvelth) [reply lib/FunctionIsolation/InvokeIsolatedFunctions.cpp]: Ah, my bad, this was a question to myself - "will I ever see this assert trigger or not?". But then I never saw it - and forgot about it :D

My bad, this should have been accompanied with a `WIP`.

Comment (Cvelth) [reply lib/FunctionIsolation/InvokeIsolatedFunctions.cpp]: Either way, I removed the message, just kept the assert, as I feel like it makes sense in the context even without extra words (check nothing is assigned to an object we're about to assign to).

File: lib/FunctionIsolation/InvokeIsolatedFunctions.cpp (hunk 24 above)
Comment (Cvelth) [reply lib/FunctionIsolation/InvokeIsolatedFunctions.cpp:173]: I looked around and it looks like this is safe to do as long as both modules share the same context. Which we do here, I assume. Not to mention the two modules will be merged together right after this!

File: lib/Model/Importer/Binary/ELFImporter.cpp (hunk 15 above)
Comment (Cvelth) [reply lib/Model/Importer/Binary/ELFImporter.cpp]: Done.

File: include/revng/ADT/UpcastablePointer.h (hunk 6 above)
Comment (aleclearmind) [reply include/revng/ADT/UpcastablePointer.h]: > That being said, if you prefer, I wouldn't mind

:+1:

File: include/revng/ADT/UpcastablePointer.h [hunk 25]
```diff
@@ -52,6 +52,15 @@ concept UpcastablePointerLike = Dereferenceable<T> and Upcastable<pointee<T>>;
...
 
+/// This is an upcasting helper `UpcastablePointer` below takes advantage of
+/// heavily. When possible, prefer the member function instead.
+///
+/// That being said, this free function is not tied to the object and can be
+/// used on normal pointers, as long as LLVM-style RTTI conventions
+/// are adhered to.
+///
+/// \note this also correctly handles `llvm::Error` and `llvm::Expected`
+/// when they are passed into \param IfNull.
```
Comment (aleclearmind) [inline include/revng/ADT/UpcastablePointer.h]: What does it mean to "handle correctly"? Be explicit.

> \note `IfNull` can be a `llvm::Error`. In such case, in case of successful upcast, `IfNull` will be consumed.

File: include/revng/ADT/UpcastablePointer.h [hunk 26]
```diff
@@ -52,6 +52,15 @@ concept UpcastablePointerLike = Dereferenceable<T> and Upcastable<pointee<T>>;
 template<typename T>
 concept NotUpcastablePointerLike = not UpcastablePointerLike<T>;
 
+/// This is an upcasting helper `UpcastablePointer` below takes advantage of
+/// heavily. When possible, prefer the member function instead.
```
Comment (aleclearmind) [inline include/revng/ADT/UpcastablePointer.h]: I think a comma is missing or something.

File: lib/Model/Importer/Binary/MachOImporter.cpp (hunk 8 above)
Comment (aleclearmind) [reply lib/Model/Importer/Binary/MachOImporter.cpp]: Mmmh, I don't understand. Can you make an example?

File: tests/unit/Pipeline.cpp (hunk 11 above)
Comment (aleclearmind) [reply tests/unit/Pipeline.cpp]: Let's drop it. We'll leave boost test one day.

File: lib/Model/Importer/Binary/ELFImporter.cpp (hunk 15 above)
Comment (aleclearmind) [reply lib/Model/Importer/Binary/ELFImporter.cpp]: Please mark as solved, if you don't want me to take a look.

File: lib/Pipeline/LLVMContainer.cpp (hunk 17 above)
Comment (aleclearmind) [reply lib/Pipeline/LLVMContainer.cpp]: `s/Untracted/Untracked/`

File: libexec/revng/check-conventions [hunk 27]
```diff
@@ -633,6 +633,12 @@ class CheckCommitPass(Pass):
         if len(subject) > 50:
             problems.append("Commit subject line exceeds 50 characters")
 
+        if "wip" in subject.lower():
```
Comment (aleclearmind) [inline libexec/revng/check-conventions]: Maybe I'd go with WIP or use a regexp and do `\bwip\b`. Otherwise we ban "wipe" `:D`

File: lib/Model/Importer/DebugInfo/DwarfImporter.cpp [hunk 28]
```diff
@@ -1254,6 +1254,7 @@ void DwarfImporter::import(StringRef FileName, const ImporterOptions &Options) {
 
       auto DebugFilePath = findDebugInfoFileByName(FileName, DebugFile, ELF);
       if (!DebugFilePath) {
+        // TODO: Should this not be an assert?
```
Comment (aleclearmind) [inline lib/Model/Importer/DebugInfo/DwarfImporter.cpp]: Agreed

File: lib/Model/Importer/Binary/MachOImporter.cpp (hunk 8 above)
Comment (Cvelth) [reply lib/Model/Importer/Binary/MachOImporter.cpp]: I tried to make a godbolt example, but got it looks like it only works with ABI-breaking-checks disabled. And the assert we're interested in is hidden behind those.

So, the best I can do is paste some code from LLVM:
```cpp
// default constructor:
  /// Create a success value. Prefer using 'Error::success()' for readability
  Error() {
    setPtr(nullptr);
    setChecked(false);
  }

// is called from:
class ErrorSuccess final : public Error {};

inline ErrorSuccess Error::success() { return ErrorSuccess(); }
```

Note the `setChecked(false)` part. Which is then asserted on in the destructor:
```cpp
  ~Error() {
    assertIsChecked();
    delete getPtr();
  }

  void assertIsChecked() {
#if LLVM_ENABLE_ABI_BREAKING_CHECKS
    if (LLVM_UNLIKELY(!getChecked() || getPtr()))
      fatalUncheckedError();
#endif
  }
```

Which means:
```cpp
{
  Error TheError = Error::success();
}
```
would assert because `TheError` is unchecked.

File: tests/unit/Pipeline.cpp (hunk 11 above)
Comment (Cvelth) [reply tests/unit/Pipeline.cpp]: Leave it for what? I'd assume most decent testing framework have something like that.

File: lib/Model/Importer/Binary/ELFImporter.cpp (hunk 15 above)
Comment (Cvelth) [reply lib/Model/Importer/Binary/ELFImporter.cpp]: Ah, my bad. I blame github because I'm sure I clicked the button (unless I missed or something) :P

File: lib/Pipeline/LLVMContainer.cpp (hunk 17 above)
Comment (Cvelth) [reply lib/Pipeline/LLVMContainer.cpp]: Ah! My bad.

File: include/revng/ADT/UpcastablePointer.h (hunk 26 above)
Comment (Cvelth) [reply include/revng/ADT/UpcastablePointer.h]: Right, good catch:
```cpp
/// This is an upcasting helper, `UpcastablePointer` below takes advantage of
/// heavily. When possible, prefer the member function instead.
```

File: libexec/revng/check-conventions (hunk 27 above)
Comment (Cvelth) [reply libexec/revng/check-conventions]: Who needs wiping anyway? /j

Jokes aside, yeah, my bad. It looks like it's literally just `wipe` and `swipe` (also `swip[p]le`, but I have never heard that one :P), but let's go with the regex.

Comment (Cvelth) [reply libexec/revng/check-conventions]: Speaking about which!!! Is there a reason why we don't check this in the CI?

I can see why we don't check the compilability, as that takes a while and, sometimes, for nastier PRs we actually want an ability to break it, but conventions?

I feel like we should really consider running the [top one](https://github.com/revng/orchestra/blob/develop/.orchestra/config/components/revng.yml#L57) before merging branches!

Comment (aleclearmind) [reply libexec/revng/check-conventions]: We have:

```yaml
post_install_script: |
  if [ "$RUN_TESTS" -eq 1 ]; then
    cd "$SOURCE_DIR"
    ./libexec/revng/check-conventions
  fi
```

The CI runs `orc install revng --test`.

File: tests/unit/Pipeline.cpp (hunk 11 above)
Comment (aleclearmind) [reply tests/unit/Pipeline.cpp]: I was thinking that it restart after aborts, but here it won't simply abort.

To be honest I don't think people usually write tests thinking that after a failure things continue to run and it's kinda hard to enforce this. I'm not user I want to pollute the tests with these `if`s. :thinking:

Have you ever found it useful? It's nice in theory, but did it ever helped you?
I usually fix one failure a time.

File: lib/Model/Importer/Binary/MachOImporter.cpp (hunk 8 above)
Comment (aleclearmind) [reply lib/Model/Importer/Binary/MachOImporter.cpp]: Ah I see. But the doc of `bindTable` does not suggest to do that.
Other users in LLVM do not do that.
What matters is that eventually we do check the error.

File: libexec/revng/check-conventions (hunk 27 above)
Comment (Cvelth) [reply libexec/revng/check-conventions]: Ah, so it's only `--check-commit` that's missing! Should we add it?

Comment (Cvelth) [reply libexec/revng/check-conventions]: Ah, wait, no, that needs to be run on every commit! So yeah, I still feel like we should run something `check-branch`-like in CI.

File: tests/unit/Pipeline.cpp (hunk 11 above)
Comment (Cvelth) [reply tests/unit/Pipeline.cpp]: Honestly, I don't think I even interacted with unit-tests to begin with :P

The failures we get from those are usually well telegraphed, so I don't think I ever actually needed to "debug" those in a traditional sense. Delta attaching the debugger and seeing which exact assert triggers.

So I don't have a strong stance on this in the first place. It's the matter of being technically-correct. But again, those are just tests, so who cares.

File: libexec/revng/check-conventions (hunk 27 above)
Comment (aleclearmind) [reply libexec/revng/check-conventions]: Eeeh eventually. Not today.

File: lib/Model/Importer/Binary/MachOImporter.cpp (hunk 8 above)
Comment (Cvelth) [reply lib/Model/Importer/Binary/MachOImporter.cpp]: I guess this is indeed hair-splitting. Because if there are no conditions - you're going to see the fact that this empty error is unchecked the very first time you run this. So maybe these aren't important after all.

File: include/revng/ADT/UpcastablePointer.h (hunk 26 above)
Comment (Cvelth) [reply include/revng/ADT/UpcastablePointer.h]: (please don't close threads like these, or might miss them when doing the final pass and the real comment will be left without its comma :P)

File: lib/Model/Importer/Binary/MachOImporter.cpp (hunk 8 above)
Comment (aleclearmind) [reply lib/Model/Importer/Binary/MachOImporter.cpp]: Yeah, I'm just trying to figure out if this is a good practice we should extend elsewhere as well. 

If not, let's leave it out.

File: tests/unit/Pipeline.cpp (hunk 11 above)
Comment (aleclearmind) [reply tests/unit/Pipeline.cpp]: Your call.

File: include/revng/ADT/UpcastablePointer.h (hunk 26 above)
Comment (aleclearmind) [reply include/revng/ADT/UpcastablePointer.h]: Ah OK, I thought you reporting it meant it was already in the codebase.

Comment (Cvelth) [reply include/revng/ADT/UpcastablePointer.h]: I usually explicitly use a word "done" or "fixed" in that case!

File: include/revng/ADT/UpcastablePointer.h (hunk 6 above)
Comment (Cvelth) [reply include/revng/ADT/UpcastablePointer.h]: Done.

File: lib/Model/Importer/Binary/MachOImporter.cpp (hunk 8 above)
Comment (Cvelth) [reply lib/Model/Importer/Binary/MachOImporter.cpp]: Gone.

File: tests/unit/Pipeline.cpp (hunk 11 above)
Comment (Cvelth) [reply tests/unit/Pipeline.cpp]: I dropped this. Let's just drop the pipeline and be done with these tests!

File: lib/Pipeline/LLVMContainer.cpp (hunk 17 above)
Comment (Cvelth) [reply lib/Pipeline/LLVMContainer.cpp]: Fixed.

File: include/revng/ADT/UpcastablePointer.h (hunk 25 above)
Comment (Cvelth) [reply include/revng/ADT/UpcastablePointer.h]: Done.

File: include/revng/ADT/UpcastablePointer.h (hunk 26 above)
Comment (Cvelth) [reply include/revng/ADT/UpcastablePointer.h]: Done.

File: libexec/revng/check-conventions (hunk 27 above)
Comment (Cvelth) [reply libexec/revng/check-conventions]: Done (`\bwip\b`).

File: lib/Model/Importer/DebugInfo/DwarfImporter.cpp (hunk 28 above)
Comment (Cvelth) [reply lib/Model/Importer/DebugInfo/DwarfImporter.cpp]: Done.
